sdk/src/mesh_dp.c

```diff
@@ -0,0 +1,779 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <pthread.h>
+#include "mesh_dp_internal.h"
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <errno.h>
+#include <bsd/string.h>
+#include <stdatomic.h>
+#include <unistd.h>
+
+/**
+ * Isolation interface for testability. Accessed from unit tests only.
+ */
+struct mesh_internal_ops_t mesh_internal_ops = {
+    .create_conn = mcm_create_connection,
+    .destroy_conn = mcm_destroy_connection,
+    .dequeue_buf = mcm_dequeue_buffer,
+    .enqueue_buf = mcm_enqueue_buffer,
+};
+
+/**
+ * Connection list head type definition
+ */
+LIST_HEAD(MeshConnectionListHead, MeshConnectionContext);
+
+/**
+ * Mesh client context structure
+ */
+typedef struct MeshClientContext {
+    MeshClientConfig config;
+
+    struct MeshConnectionListHead conn_list_head;
+    pthread_mutex_t mx;
+
+    atomic_int conn_num;
+
+} MeshClientContext;
+
+/**
+ * Mesh connection buffer structure
+ */
+typedef struct MeshBufferContext {
+    /**
+     * NOTE: The __public structure is directly mapped in the memory to the
+     * MeshBuffer structure, which is publicly accessible to the user.
+     * Therefore, the __public structure _MUST_ be placed first here.
+     */
+    MeshBuffer __public;
+
+    /**
+     * NOTE: All declarations below this point are hidden from the user.
+     */
+    mcm_buffer *buf;
+
+} MeshBufferContext;
+
+/**
+ * Create a new mesh client
+ */
+int mesh_create_client(MeshClient **mc, MeshClientConfig *cfg)
+{
+    MeshClientContext *mc_ctx;
+    int err;
+
+    if (!mc)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    mc_ctx = mesh_calloc(sizeof(MeshClientContext));
+    if (!mc_ctx) {
+        *mc = NULL;
+        return -ENOMEM;
+    }
+
+    atomic_init(&mc_ctx->conn_num, 0);
+
+    err = pthread_mutex_init(&mc_ctx->mx, NULL);
+    if (err) {
+        mesh_free(mc_ctx);
+        *mc = NULL;
+        return err;
+    }
+
+    LIST_INIT(&mc_ctx->conn_list_head);
+
+    if (cfg)
+        mesh_memcpy(&mc_ctx->config, cfg, sizeof(MeshClientConfig));
+
+    if (mc_ctx->config.max_conn_num == 0)
+        mc_ctx->config.max_conn_num = MESH_CLIENT_DEFAULT_MAX_CONN;
+
+    if (mc_ctx->config.timeout_ms == 0)
+        mc_ctx->config.timeout_ms = MESH_CLIENT_DEFAULT_TIMEOUT_MS;
+
+    *mc = (MeshClient *)mc_ctx;
+
+    return 0;
+}
+
+/**
+ * Delete mesh client
+ */
+int mesh_delete_client(MeshClient **mc)
+{
+    MeshClientContext *mc_ctx;
+
+    if (!mc)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    mc_ctx = (MeshClientContext *)(*mc);
+
+    if (!mc_ctx)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    pthread_mutex_destroy(&mc_ctx->mx);
+
+    /**
+     * TODO: Shutdown and deallocate connections here
```

Comment:
It is the user's responsibility to call `mesh_connection_delete()` to deallocate the connection. The user had to do that in the previous PR, and I'm not changing this flow.  This `TODO` has been left here for the future to remind us to implement the automatic connections shutdown and deallocation in `mesh_delete_client()` even when the user forgets to do that. So far it can't be implemented here because it requires additional mechanisms to gracefully shutdown the connection and synchronize everything. This is a future task.

Reply:
So currently we will have a memory leak if the user don't use mesh_delete_client() for every connection?

Is it really that complicated to implement this feature? It seems that traversing though the list and calling something like mesh_delete_client on every node will be sufficient.

Reply:
No, the user must call `mesh_delete_client()` only once after deleting all connections with `mesh_delete_connection()`.

I believe we don't want the user to call `mesh_delete_client()` while the connection is still active in a parallel thread receiving or sending data. We don't want that to happen because there are no mechanisms of shutdown synchronization between the client and the connection. Before deleting a connection, the client must raise a signal to notify other threads that the connection and its resources are about to destroy. Then, every thread should react to this notification by closing the connection and decreasing the refcount. After that, when the connection refcount is decreased down to zero, the client can finally delete the connection.

This sort of a mechanism could be developed in the future as a new feature. For now, it's the user's responsibility to delete all connections before deleting the client.

Reply:
Fixed in eae8b083e0210567994eb379b5e228742451c44b.